01-data-structures/05-hashes-part-2/open_addressing/open_addressing.rb

```diff
@@ -2,29 +2,96 @@
 
 class OpenAddressing
   def initialize(size)
+    @items = Array.new(size)
+    @entries = 0
   end
 
   def []=(key, value)
+    node = Node.new(key, value)
+    i = index(key, self.size)
+    if @items[i].nil?
+      @items[i] = node
+    elsif @items[i].key == key
+      @items[i].value = value
+      if self.next_open_index(i) == -1
+        resize
+      end
+    else
+      j = self.next_open_index(i)
+      if j != -1
+        @items[j] = node
+      else
+        resize
+        self[key] = value
+      end
+    end
   end
 
   def [](key)
+    (0...size()).each do |i|
+      if @items[i] != nil
+        if @items[i].key == key
+          return @items[i].value
+        end
+      end
+    end
   end
 
   # Returns a unique, deterministically reproducible index into an array
   # We are hashing based on strings, let's use the ascii value of each string as
   # a starting point.
   def index(key, size)
+    key.sum % size
   end
 
   # Given an index, find the next open index in @items
   def next_open_index(index)
+    i = index
+    for i in index..self.size-1 do
+      if @items[i].nil?
+        return i
+      end
+    end
+    if index > 0
```

Comment:
The open indexing algorithm does not need to circle around. Once it hits the end, that is it. This correction will affect your implementation of `[](key)` as well.